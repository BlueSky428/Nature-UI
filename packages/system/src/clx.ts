type Other = string | number | null | boolean | undefined;
export type ClassValue = ClassArray | ClassDict | Other;

interface ClassDict {
  [key: string]: Other;
}
export type ClassArray = Array<ClassValue>;

const clx = (...classNames: ClassValue[]): string => classNames.filter(Boolean).join(' ');

export default clx;
