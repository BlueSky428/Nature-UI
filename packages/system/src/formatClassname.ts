export const formatClassNames = (className: string): string => {
  const withCss = className.split(' ').filter((v) => v.match(/css-\w+/));
  const ArrayClassName = className
    .split(' ')
    .filter((v) => !v.match(/css-\w+/));

  const result = {};
  const required: string[] = [];
  ArrayClassName.forEach((v) => {
    const splitCn = v.split('-');

    if (splitCn.length === 1) {
      required.push(v);
    } else if (isNaN(parseInt(splitCn[1]))) {
      required.push(v);
    } else {
      const firstChar = splitCn[0];
      result[firstChar] = v;
    }
  });

  const cn = `${Object.values(result).join(' ')} ${required.join(
    ' ',
  )} ${withCss?.join(' ')}`;

  return cn;
};
