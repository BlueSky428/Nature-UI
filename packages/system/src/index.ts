export { css, keyframes } from '@emotion/css';
export * from './jsx';
export * from './system';
export * from './system.types';
export * from './system.utils';
