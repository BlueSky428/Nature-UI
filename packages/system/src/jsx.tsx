import * as React from 'react';
import { Dict } from '@nature-ui/utils';
import { jsx as emotion } from '@emotion/core';
/*
 * export const jsx = (
 *   type: React.ElementType = 'div',
 *   props: Dict,
 *   ...children: React.ReactNode[]
 * ) => React.createElement(type, props, ...children);
 */

export const jsx = (
  type: React.ElementType = 'div',
  props: Dict,
  ...children: React.ReactNode[]
) => emotion.apply(undefined, [type, props, ...children]);

export default jsx;
