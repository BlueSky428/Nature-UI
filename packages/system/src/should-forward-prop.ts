export interface ValidHTMLProps {
  htmlWidth?: string | number;
  htmlHeight?: string | number;
  htmlSize?: string | number;
}
