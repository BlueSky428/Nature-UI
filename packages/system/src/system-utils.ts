import * as React from 'react';
import { isString, UnionStringArray, __DEV__ } from '@nature-ui/utils';

import { ForwardRefComponent } from './system-types';
import { pseudoSelectors } from './pseudo';

/**
 * Carefully selected html elements for nature components.
 * This is mostly for `nature.<element>` syntax.
 */
export const domElements = [
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'b',
  'bdi',
  'bdo',
  'big',
  'blockquote',
  'button',
  'caption',
  'cite',
  'circle',
  'code',
  'col',
  'dd',
  'del',
  'details',
  'dfn',
  'div',
  'dl',
  'dt',
  'em',
  'fieldset',
  'figcaption',
  'figure',
  'footer',
  'form',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'header',
  'hr',
  'i',
  'img',
  'input',
  'ins',
  'kbd',
  'label',
  'legend',
  'li',
  'main',
  'mark',
  'nav',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'path',
  'picture',
  'pre',
  'q',
  'rect',
  's',
  'svg',
  'section',
  'select',
  'small',
  'span',
  'strong',
  'sub',
  'summary',
  'sup',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'tr',
  'u',
  'ul',
  'video',
] as const;

export type DOMElements = UnionStringArray<typeof domElements>;

export const pseudoProps = (props: any) => {
  let result = {};

  Object.keys(props).forEach((prop) => {
    if (prop in pseudoSelectors) {
      const style = { [prop]: props[prop] };

      result = {
        ...result,
        ...style,
      };
    }
  });

  return result;
};

export const isTag = (target: any) => {
  return isString(target) && __DEV__
    ? target.charAt(0) === target.charAt(0).toLowerCase()
    : true;
};

export const getComponentName = (primitive: React.ComponentType | string) => {
  return (
    (__DEV__ ? isString(primitive) && primitive : false) ||
    (!isString(primitive) && primitive.displayName) ||
    (!isString(primitive) && primitive.name) ||
    'NatureComponent'
  );
};

export const getDisplayName = (primitive: any) => {
  return isTag(primitive) ? `nature.${primitive}` : getComponentName(primitive);
};

export const forwardRef = <P>(
  comp: (props: P, ref: React.Ref<any>) => React.ReactElement | null,
) => {
  return (React.forwardRef(comp as any) as unknown) as ForwardRefComponent<P>;
};
