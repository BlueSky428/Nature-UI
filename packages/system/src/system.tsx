import * as React from 'react';
import { As, runIfFn } from '@nature-ui/utils';

const createComponent = <T extends As>(component: T) => {
  return (...interpolations: any[]) => {
    const Component = React.forwardRef(
      ({ as, ...props }: any, ref: React.Ref<any>) => {
        interpolations.forEach((interpolation) => {
          runIfFn(interpolation, {});
        });
      }
    );

    Component.displayName = getDisplayName(component);
    Component.defaultProps = (component as any).defaultProps;
  };
};

function styled<T extends As>(component: T) {
  return;
}
