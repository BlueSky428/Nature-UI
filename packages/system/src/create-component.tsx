import * as React from 'react';
import {
  As,
  Dict,
  isEmptyObject,
  isString,
  isUndefined,
} from '@nature-ui/utils';
import hoist from 'hoist-non-react-statics';

import { jsx } from './jsx';
import { getDisplayName } from './system-utils';
import { NatureComponent } from './system-types';

export const createComponent = <T extends As>(component: T) => {
  // return (...interpolations: any[]) => {
  const Component = React.forwardRef(
    ({ as, ...props }: any, ref: React.Ref<any>) => {
      /*
       * interpolations.forEach((interpolation) => {
       *   runIfFn(interpolation, {});
       * });
       */

      const element = as || component;

      const isTag = isString(element);

      const computedProps: Dict = isTag ? { ...props } : { ...props };
      // const computedProps: Dict = !isTag && { ...props };

      if (isEmptyObject(computedProps.css) || isUndefined(computedProps.css)) {
        delete computedProps.css;
      }

      return jsx(element, {
        ref,
        ...computedProps,
      });
    }
  );

  Component.displayName = getDisplayName(component);
  Component.defaultProps = (component as any).defaultProps;

  // hoist all non-react statics attached to the `component` prop
  const MainComponent = hoist(Component, component as any);

  return MainComponent as NatureComponent<T>;
  // };
};
