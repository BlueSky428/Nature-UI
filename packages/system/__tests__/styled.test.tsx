/** @jsx jsx */
import { render } from '@nature-ui/test-utils';

import { nature, jsx } from '../src';

describe('@nature-ui/system', () => {
  test('as jsx element', () => {
    const Div = nature('div');
    const tools = render(<Div as='span'>a</Div>);

    expect(tools.getByAltText('a')).toMatchInlineSnapshot(`
    <span>
    a
    </span>
    `);
  });
});
