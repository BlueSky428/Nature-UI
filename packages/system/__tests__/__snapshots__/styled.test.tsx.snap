// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/system it allows pass through props 1`] = `
<DocumentFragment>
  <img
    data-testid="img"
    width="60px"
  />
</DocumentFragment>
`;

exports[`@nature-ui/system nature elements - renders with tailwind className props 1`] = `
<DocumentFragment>
  <div
    class="mt-40"
  >
    Welcome
  </div>
</DocumentFragment>
`;
